Use POSSUM database along with Google spreadsheets

riniangreani opened the pull request:
**For POSSUM Status sheet:**
- As discussed, this sheet is used by the general public to check for statuses, therefore we need to keep it updated by the script.
- Database tables are updated along with sheet update. 
- Additional columns in **observation** table: single_SB_1D_pipeline_band1, single_SB_1D_pipeline_band2
- Additional columns in **tile** table: 3d_pipeline_band1_status, 3d_pipeline_band2_status

**For POSSUM Validation sheet:**
- Partial Tile Pipeline - regions - Band 1: replaced with database table (**partial_tile_1d_pipeline**) and additional columns in **observation** table (1d_pipeline_validation_band1 and 1d_pipeline_validation_band2).
- Survey Tiles - Band 1, Survey Tiles - Band 2: left as is due to manual entries required in the validation process.

Unit tests will be done in a separate PR.

riniangreani committed:
Add script to populate partial_tile_1d_pipeline in the database (to replace possum validation sheet for partial tile pipeline)

riniangreani committed:
Added script to update and populate observation and tile tables, some 1d pipeline handling

riniangreani committed:
Update calls to update Google sheets to database

riniangreani committed:
Formatting changes

riniangreani committed:
Cleaning up

riniangreani committed:
More tidying up

riniangreani committed:
Add unique constraint. More tidying up

riniangreani commented:
Based on Erik's comments, the following changes need to be made:
1. Unique constraint for partial_tile_1d_pipeline table  does not need "number_sources". 
Additional rules from Erik (nice to have):
- Every row is unique, even when only considering the columns field_name	sbid	tile1	tile2	tile3	tile4
- No row has tileX == tileY where X!=Y
- if 4 tile numbers are present, type == corner or type==corner - crosses projection boundary!
- if 2 tile numbers are present (the first two are not null, the last two are null), type==edge or type==edge - crosses projection boundary!
- if 1 tile number is present type == center

2. The database needs to be updated before each run as the sheet changes. We need to ensure database insertion script is always updated (that means the partial tile table needs to be emptied and reinserted as the row index etc may change along with the values). 

3. Split partial tile table into 2 for band 1 and band 2, since the footprints are different per band. 

4. DB update to 3D columns are needed for: 
- log_processing_status_3Dpipeline.py -> update_validation_sheet
- ingest3dpipeline.py  -> update_validation_sheet
- check_ingest_3dpipeline.py -> get_tiles_for_ingest, update_status
- check if we missed any other 3D updates

riniangreani committed:
Merge branch 'main' into prefect_prototype

riniangreani committed:
Split partial_tile_1d_pipeline into 2 tables for band 1 and band 2

riniangreani committed:
Create observation_1d_pipeline_band1 and observation_1d_pipeline_band2 tables

riniangreani committed:
Remove unused code

riniangreani committed:
Added more checks for tiles based on Erik's suggestions

riniangreani committed:
Use upserts instead of inserts because we need to keep updating the database from google sheets

riniangreani committed:
Fix up insert queries to make sure partial tiles values do not get overwritten by spreadsheet on update

riniangreani committed:
Fix up 3D database update

riniangreani committed:
implement get tiles queries

riniangreani committed:
Update hardcoded Google spreadsheets and tokens with the env file

riniangreani committed:
Fix up query

riniangreani committed:
Make queries work with django

riniangreani committed:
Add unit tests

riniangreani committed:
Added unit test for launching 1d pipeline partial tiles

riniangreani committed:
Clean up test file

riniangreani committed:
Test empty tile numbers

riniangreani committed:
Add unit test for launching 3d pipeline

riniangreani committed:
Remove unused import

riniangreani committed:
Added test for ingest3dpipeline.py

riniangreani committed:
Added test for log_processing_status_1d_partialtiles_summary, general cleanup

riniangreani committed:
Add test for log_processing_status_status_3dpipeline

riniangreani committed:
Add more unit tests for logging status

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+        SET "3d_pipeline_ingest" = %s
+        WHERE tile_id = %s;
+    """
+    params = (status, tile_number)
```
instead of having the `status`' and `tile_number` as a tuple `params`, isn't it more clear to put those inside the query using the f-string? i.e. 
```
    query = f"""
        UPDATE possum.tile_3d_pipeline_band{band_number}
        SET "3d_pipeline_ingest" = {status}
        WHERE tile_id = {tile_number};
    """
```

riniangreani replied:
I updated this for the database_queries one.

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+        'port': os.getenv('DATABASE_PORT')
+    }
+
+def execute_update_query(query, conn, params=None, verbose=True):
```
Do we need a separate `execute_update_query` when we also have `execute_query`?

riniangreani replied:
It's because when you update, it returns number of rows affected. i.e. if the update fails, it returns -1. If no rows affected, it returns 0, etc. Some of the update functions have a print that tells you whether the tile could not be found, or if the update was succesful. So I check the number of rows affected to preserve the logging. 

Execute_query, however returns the dataset for select. I mean we could use one method, but I thought it'd be cleaner when they're returning different things.

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+        SET "3d_pipeline_val" = %s, "3d_val_link" = %s
+        WHERE tile_id = %s;
+    """
+    params = (status, val_link, tile_number)
```
same question here about whether having params separately is necessary

riniangreani replied:
At first I did the changes to not use params, but then I realised it is better to use params to avoid SQL injection. I know it's probably an overmeasure for our usecase, but it's always a good practice to use SQL params anyway. It also helps with parameter type conversions e.g. string to int, vice versa. https://www.w3schools.com/sql/sql_injection.asp

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+        SET "3d_pipeline" = %s
+        WHERE tile_id = %s;
+    """
+    params = (status, tile_number)
```
same question here about whether having params separately is necessary

ErikOsinga replied:
I wonder in general if we couldn't combine some of these functions for updating the same table but just different columns into a single function such that it's easier to maintain the functions in the future?

riniangreani replied:
We could easily do that, but the downside is that the function becomes less clear and that you have to do an additional check that the column name is valid.

riniangreani replied:
Actually I just refactored a few methods into: update_3d_pipeline_table() and update_1d_pipeline_table() to be more generic with an additional column name as a parameter as suggested.

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+        ON CONFLICT (name) DO UPDATE
+        SET "1d_pipeline_validation" = %s;
+    """
+    params = (field_name, sbid, status, status)
```
same question here about whether having params separately is necessary

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+        ON CONFLICT (name) DO UPDATE
+        SET single_sb_1d_pipeline = %s;
+    """
+    params = (field_name, sbid, status, status)
```
same question here about whether having params separately is necessary + combining with function above

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+            WHERE sbid = %s AND observation = %s AND LOWER(type) like '%%crosses projection boundary%%'
+        ) AS match_found;
+    """
+    params = (sbid, observation)
```
same question here about whether having params separately is necessary

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+        WHERE ob.name = pt.observation
+        GROUP BY pt.observation, pt.sbid, ob."1d_pipeline_validation", pt."1d_pipeline";
+    """
+    return execute_query(sql, conn)
```
Looks good, but hard for me to check these kind of SQL queries in detail, would be good to chat about how I can test this!

ErikOsinga commented on automation/database_queries.py:
```diff
@@ -0,0 +1,475 @@
+    """
+    return execute_query(sql, conn)
+
+#### TEST METHODS ###
```
Could we split the tests into a separate file?

ErikOsinga commented on automation/insert_database_script.py:
I didnt check this file in detail as it seems this is only used once to create the database, right?

riniangreani replied:
Yes, only to create additional tables from the spreadsheets, and also to populate test database per unit test case.

ErikOsinga commented on automation/unit_tests/README:
Would it be in scope/possible to set up some github actions to run tests automatically whenever a new feature/commit is made on one of the "automation/" subfiles?

riniangreani replied:
It is definitely possible!

ErikOsinga commented on automation/unit_tests/check_ingest_3dpipeline_test.py:
```diff
@@ -0,0 +1,69 @@
+        db.create_test_schema(self.conn)
+        db.create_tile_3d_pipeline_tables(self.conn)
+
+        _3d_data = [
```
Could you clarify which columns this corresponds to?

riniangreani replied:
Good point, I'll add a comment. Thanks

ErikOsinga commented on automation/unit_tests/check_status_and_launch_1dpipeline_partialtiles_test.py:
```diff
@@ -0,0 +1,138 @@
+        # insert partial tile data
+        rows = [
+            # field name, sbid, tile1, tile2, tile3, tile4, type, number_sources, 1d_pipeline
+            ["EMU_1748-64", "54926", "11726", "11727", "11791", "11792", "corner - crosses projection boundary!", "", ""],
```
not necessarily anything wrong with this for testing, but I wonder if it's easier to save this as a .csv file inside the repo that gets loaded here so it's maybe easier to visualise/adapt when we want to add test cases?

riniangreani replied:
We could do taht

ErikOsinga commented on automation/unit_tests/check_status_and_launch_1dpipeline_partialtiles_test.py:
```diff
@@ -0,0 +1,138 @@
+        ]
+        for row in rows:
+            db.insert_row_into_partial_tile_table(row, '1', self.conn)
+        # insert observation data
```
same .csv comment here, and can you remind me what the columns are here?

riniangreani replied:
done

ErikOsinga commented on automation/unit_tests/check_status_and_launch_3dpipeline_v2_test.py:
```diff
@@ -0,0 +1,43 @@
+        self.conn = db_query.get_database_connection(test=True)
+        db.create_test_schema(self.conn)   
+        db.create_tile_3d_pipeline_tables(self.conn)
+        _3d_data = [
```
same comment with the .csv file

ErikOsinga commented on automation/unit_tests/ingest3dpipeline_test.py:
```diff
@@ -0,0 +1,63 @@
+        self.conn = db_query.get_database_connection(test=True)
+        db.create_test_schema(self.conn)
+        db.create_tile_3d_pipeline_tables(self.conn)
+        _3d_data = [
```
same comment with the .csv file

ErikOsinga commented on possum_pipeline_control/launch_1Dpipeline_PartialTiles_band1_pre_or_post.py:
```diff
@@ -3,7 +3,6 @@
 from datetime import datetime
 # from skaha.session import Session
 from canfar.sessions import Session
-from control_1D_pipeline_PartialTiles import get_open_sessions
```
Not sure why this is deleted here?

riniangreani replied:
Sorry that was an accident!

ErikOsinga reviewed (changes requested):
Hi Rini, this looks really good. Just minor comments on this really. 

Would be great to set up a time to meet so you can show me how I can test this and interface with the new POSSUM database/tables that you've created. I'll Slack you.

riniangreani committed:
Erik's feedback, tidying up

riniangreani commented:
Hi @ErikOsinga, thanks for your feedback. I addressed most of your comments -  the major one to use csv file for test data. 
For automating the unit tests with github action, I think we need a dedicated test server, which shouldn't be difficult to do, but I'd like to discuss with Austin when he returns. Please let me know if there's any other issues. Also I sent a teams meeting invite for Wednesday.

riniangreani committed:
Erik's feedback to make update table take on column name as parameter

riniangreani committed:
Erik's feedback to put params in the query string directly

riniangreani committed:
Rename test files so they get picked up by unittest discover, remove unused constant

riniangreani committed:
Added observation 1d pipeline table for band 2

riniangreani committed:
Refactor database tables based on austin's feedback

riniangreani committed:
Refactor to speed up bulk insert over prod server, also to commit/rollback in 1 transaction

riniangreani committed:
Make sure get_tiles_for_pipeline_run for 3D get the info from the DB, added unit test

riniangreani committed:
Update tile_id column to tile

riniangreani committed:
Fix observation_band_state2 populated with band1 data

riniangreani committed:
Reinstate using sql args to avoid sql injection

ErikOsinga committed:
Merge branch 'main' into prefect_prototype
